### ISSUE TITLE
problem with a recursive structure

### ISSUE THREAD

**Issue by foricee** [eyalz800/zpp_bits#63]
Hi, I define a struct which represents a tree.

```
struct Node {
    std::vector<std::pair<uint32_t, std::shared_ptr<Node>>> children;
    std::vector<uint32_t> word;
};
```
But gcc reports an error.
```
zpp_bits.h:1817:20: error: function 'serialize_many<std::shared_ptr<Node> &>' with deduced return type cannot be used before it is defined
return serialize_many(items...);
```
Does it mean that recursion is not yet supported or something like that?

**Comment by eyalz800** [comment]
@foricee read the section that mentions that serialization of null pointers is not supported. So it may not do what you want to do and the intent is to use optional_ptr or std::optional of the pointer. I will add a test case for recursion and report back

**Comment by eyalz800** [comment]
@foricee actually since it's inside a vector it should be fine and will stop when the vector inside some node is empty, I'm uploading a fix for gcc, I think I understood the issue.

**Comment by eyalz800** [comment]
@foricee just to help debug this since I don't have gcc on my local setup and need CI every time - can you try this -#64

**Comment by foricee** [comment]
> @foricee just to help debug this since I don't have gcc on my local setup and need CI every time - can you try this -#64

That works! Thanks a lot for such a quick reply!

**Comment by eyalz800** [comment]
Fixed in v4.3.1
